modules/adoceanBidAdapter.js

```diff
@@ -0,0 +1,192 @@
+import { _each, isStr, isArray, parseSizesInput } from '../src/utils.js';
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+import {BANNER, VIDEO} from '../src/mediaTypes.js';
+
+const BIDDER_CODE = 'adocean';
+const URL_SAFE_FIELDS = {
+  schain: true,
+  slaves: true
+};
+
+function buildEndpointUrl(emiter, payloadMap) {
+  const payload = [];
+  _each(payloadMap, function(v, k) {
+    payload.push(k + '=' + (URL_SAFE_FIELDS[k] ? v : encodeURIComponent(v)));
+  });
+
+  const randomizedPart = Math.random().toString().slice(2);
+  return 'https://' + emiter + '/_' + randomizedPart + '/ad.json?' + payload.join('&');
+}
+
+function buildRequest(bid, gdprConsent) {
+  const emiter = bid.params.emiter;
+  const masterId = bid.params.masterId;
+  const slaveId = bid.params.slaveId;
+  const payload = {
+    id: masterId,
+    slaves: ""
+  };
+  if (gdprConsent) {
+    payload.gdpr_consent = gdprConsent.consentString || undefined;
+    payload.gdpr = gdprConsent.gdprApplies ? 1 : 0;
+  }
+  if (bid.schain) {
+    payload.schain = serializeSupplyChain(bid.schain);
+  }
+
+  if (bid.userId && bid.userId.gemiusId) {
+    payload.aouserid = bid.userId.gemiusId;
+  }
+
+  const bidIdMap = {};
+  const uniquePartLength = 10;
+
+  const rawSlaveId = bid.params.slaveId.replace('adocean', '');
+  payload.slaves = rawSlaveId.slice(-uniquePartLength);
+
+  bidIdMap[slaveId] = bid.bidId;
+
+  if (bid.mediaTypes.video) {
+    if (bid.mediaTypes.video.context === 'instream') {
+      if (bid.mediaTypes.video.maxduration) {
+        payload.dur = bid.mediaTypes.video.maxduration;
+        payload.maxdur = bid.mediaTypes.video.maxduration;
+      }
+      if (bid.mediaTypes.video.minduration) {
+        payload.mindur = bid.mediaTypes.video.minduration;
+      }
+      payload.spots = 1;
+    }
+    if (bid.mediaTypes.video.context === 'adpod') {
+      const durationRangeSec = bid.mediaTypes.video.durationRangeSec;
+      if (!bid.mediaTypes.video.adPodDurationSec || !isArray(durationRangeSec) || durationRangeSec.length === 0) {
+        return;
+      }
+      const spots = getAdPodSpots(bid.mediaTypes.video.adPodDurationSec, bid.mediaTypes.video.durationRangeSec);
+      const maxDuration = Math.max(...durationRangeSec);
+      payload.dur = bid.mediaTypes.video.adPodDurationSec;
+      payload.maxdur = maxDuration;
+      payload.spots = spots;
+    }
+  } else if (bid.mediaTypes.banner) {
+    payload.aosize = parseSizesInput(bid.mediaTypes.banner.sizes).join(',');
+  }
+
+  return {
+    method: 'GET',
+    url: buildEndpointUrl(emiter, payload),
+    data: '',
+    bidIdMap: bidIdMap
+  };
+}
+
+function getAdPodSpots(adPodDurationSec, durationRangeSec) {
+  const minAllowedDuration = Math.min(...durationRangeSec);
+  const numberOfSpots = Math.floor(adPodDurationSec / minAllowedDuration);
+  return numberOfSpots;
+}
+
+const SCHAIN_FIELDS = ['asi', 'sid', 'hp', 'rid', 'name', 'domain', 'ext'];
+function serializeSupplyChain(schain) {
```

Comment:
Jak to idzie tylko do AO jako adservera to nie do końca wiem do czego to będzie wykorzystywane